File: backend/src/apps/organizations/models/organization.py
from django.db import models
from django.contrib.auth import get_user_model

User = get_user_model()


class Organization(models.Model):
    """
    Организация для совместной работы
    """
    
    name = models.CharField("Название", max_length=255)
    description = models.TextField("Описание", blank=True, null=True)
    owner = models.ForeignKey(
        User,
        on_delete=models.CASCADE,
        verbose_name="Владелец",
        related_name="owned_organizations"
    )
    created_at = models.DateTimeField("Создано", auto_now_add=True)
    updated_at = models.DateTimeField("Обновлено", auto_now=True)
    is_active = models.BooleanField("Активна", default=True)
    
    class Meta:
        db_table = "organizations"
        verbose_name = "Организация"
        verbose_name_plural = "Организации"
        ordering = ['-created_at']
    
    def __str__(self):
        return self.name


class OrganizationMember(models.Model):
    """
    Член организации с ролью
    """
    
    class RoleChoices(models.TextChoices):
        ADMIN = 'admin', 'Администратор'
        MEMBER = 'member', 'Участник'
        VIEWER = 'viewer', 'Зритель'
    
    organization = models.ForeignKey(
        Organization,
        on_delete=models.CASCADE,
        verbose_name="Организация",
        related_name="members"
    )
    user = models.ForeignKey(
        User,
        on_delete=models.CASCADE,
        verbose_name="Пользователь",
        related_name="organizations"
    )
    role = models.CharField(
        "Роль",
        max_length=20,
        choices=RoleChoices.choices,
        default=RoleChoices.MEMBER
    )
    joined_at = models.DateTimeField("Присоединился", auto_now_add=True)
    
    class Meta:
        db_table = "organization_members"
        verbose_name = "Член организации"
        verbose_name_plural = "Члены организации"
        unique_together = ['organization', 'user']
        ordering = ['-joined_at']
    
    def __str__(self):
        return f"{self.user.username} - {self.role} in {self.organization.name}"

File: frontend/src/types/Organization.ts
// Organization types

export interface Organization {
  id: number;
  name: string;
  description: string | null;
  owner_id: number;
  created_at: string;
  updated_at: string;
  is_active: boolean;
}

export interface OrganizationCreate {
  name: string;
  description?: string;
}

export interface OrganizationMember {
  id: number;
  user_id: number;
  username: string;
  role: 'admin' | 'member' | 'viewer';
  joined_at: string;
}

export interface Vault {
  id: number;
  name: string;
  description: string | null;
  organization_id: number;
  created_by_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface VaultCreate {
  name: string;
  description?: string;
  organization_id: number;
}

export interface VaultAccess {
  id: number;
  vault_id: number;
  user_id: number;
  username: string;
  permission: 'read' | 'write' | 'admin';
  granted_at: string;
}

export interface VaultGrantAccess {
  user_id: number;
  permission: 'read' | 'write' | 'admin';
}
